ern/index_test.go

```diff
@@ -0,0 +1,197 @@
+	go func() {
+		defer close(stream)
+		for _, cid := range cids {
+			stream <- cid
+		}
+	}()
+
+	// create a test SQLite3 db
+	tmpDir, err := ioutil.TempDir("", "musicbrainz-index-test")
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer os.RemoveAll(tmpDir)
+	db, err := sql.Open("sqlite3", filepath.Join(tmpDir, "index.db"))
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer db.Close()
+
+	// index the stream of ERNs
+	indexer, err := NewIndexer(db, store)
+	if err != nil {
+		t.Fatal(err)
+	}
+	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
+	defer cancel()
+	if err := indexer.Index(ctx, stream); err != nil {
+		t.Fatal(err)
+	}
+
+	// check the MessageSender and MessageRecipient were indexed into the
+	// party table
+	for _, partyID := range []string{"DPID_OF_THE_SENDER", "DPID_OF_THE_RECIPIENT"} {
+		rows, err := db.Query(`SELECT cid FROM party WHERE id = ?`, partyID)
+		if err != nil {
+			t.Fatal(err)
+		}
+		defer rows.Close()
+		var id string
+		for rows.Next() {
+			// if we've already set id then we have a duplicate
+			if id != "" {
+				t.Fatalf("duplicate entries for PartyId %q", partyID)
+			}
+			if err := rows.Scan(&id); err != nil {
+				t.Fatal(err)
+			}
+
+			// check we can get the object from the store
+			cid, err := cid.Parse(id)
+			if err != nil {
+				t.Fatal(err)
+			}
+			obj, err := store.Get(cid)
+			if err != nil {
+				t.Fatal(err)
+			}
+
+			// check the object has the correct PartyId
+			graph := meta.NewGraph(store, obj)
+			v, err := graph.Get("PartyId", "@value")
+			if err != nil {
+				t.Fatal(err)
+			}
+			actual, ok := v.(string)
+			if !ok {
+				t.Fatalf("expected PartyId value to be string, got %T", v)
+			}
+			if actual != partyID {
+				t.Fatalf("expected PartyId value %q, got %q", partyID, actual)
+			}
+		}
+
+		// check we got a result and no db errors
+		if id == "" {
+			t.Fatalf("party %q not found", partyID)
+		} else if err := rows.Err(); err != nil {
+			t.Fatal(err)
+		}
+	}
+
+	// check all the ERNs were indexed into the ern table
+	for _, cid := range cids {
+		obj, err := store.Get(cid)
+		if err != nil {
+			t.Fatal(err)
+		}
+		graph := meta.NewGraph(store, obj)
+		var (
+			messageID string
+			threadID  string
+		)
+		row := db.QueryRow(`SELECT message_id, thread_id FROM ern WHERE cid = ?`, cid.String())
+		if err := row.Scan(&messageID, &threadID); err != nil {
+			t.Fatal(err)
+		}
+		for field, actual := range map[string]string{
+			"MessageId":       messageID,
+			"MessageThreadId": threadID,
+		} {
+			v, err := graph.Get("NewReleaseMessage", "MessageHeader", field, "@value")
+			if err != nil {
+				t.Fatal(err)
+			}
+			expected, ok := v.(string)
+			if !ok {
+				t.Fatalf("expected %s to be a string, got %T", field, v)
+			}
+			if actual != expected {
+				t.Fatalf("expected %s to be %q, got %q", field, expected, actual)
+			}
+		}
+	}
+
+	// check SoundRecording objects were indexed
+	for isrc, title := range map[string]string{
```

Comment:
Judging from the code, an `id` for a `SoundRecording` can be `ISRC`,`Catalognumber` and `ProprietaryId`. Maybe test for all?

It also seems to only need *either* "SoundRecordingId" or `ReferenceTitle`. Maybe the test should also account for that?

Reply:
Good spot, so I'm actually keen to run this on real ERNs before writing too much code around this (I actually don't think `CatalogNumber` and `ProprietaryId` are used so we may end up dropping them), so I'm going to leave the tests like this.